fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/query/FromJoin.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.database.utils.query;
+
+import com.ibm.fhir.database.utils.query.expression.StatementRenderer;
+import com.ibm.fhir.database.utils.query.expression.StringExpNodeVisitor;
+import com.ibm.fhir.database.utils.query.node.ExpNode;
+
+/**
+ * Models a JOIN element in the from clause
+ */
+public class FromJoin extends FromItem {
+
+    private final JoinType joinType;
+
+    private final ExpNode joinOnPredicate;
+
+    // An enumeration of join types
+    public enum JoinType {
+        INNER_JOIN("INNER JOIN"),
+        LEFT_OUTER_JOIN("LEFT OUTER JOIN"),
+        FULL_OUTER_JOIN("FULL OUTER JOIN");
+
+        private final String joinText;
+
+        private JoinType(String joinText) {
+            this.joinText = joinText;
+        }
+
+        @Override
+        public String toString() {
+            return this.joinText;
+        }
+    }
+
+    /**
+     * Protected constructor
+     *
+     * @param joinType
+     * @param rowSource
+     * @param alias
+     */
+    protected FromJoin(JoinType joinType, RowSource rowSource, Alias alias, ExpNode joinOnPredicate) {
+        super(rowSource, alias);
+        this.joinType = joinType;
+        this.joinOnPredicate = joinOnPredicate;
+    }
+
+    /**
+     * Protected constructor for a join which doesn't include an alias
+     *
+     * @param joinType
+     * @param rowSource
+     */
+    protected FromJoin(JoinType joinType, RowSource rowSource, ExpNode joinOnPredicate) {
+        super(rowSource, null);
+        this.joinType = joinType;
+        this.joinOnPredicate = joinOnPredicate;
+    }
+
+    @Override
+    public boolean isAnsiJoin() {
+        // We are definitely an ANSI-style join...which is why this exists in the first place
+        return true;
+    }
```

Comment:
When would this be false? e.g. the scout project?

Reply:
No, it's to handle the mix of items in the FROM clause. You can get simple table references (with the join predicate in the WHERE clause) or sub-queries, or in this case, an ANSI-style JOIN statement with an ON clause. This just allows us to mix styles and generate the correct statement.